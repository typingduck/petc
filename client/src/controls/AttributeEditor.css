
.petc-attr-editor {
  padding: 20px 20px 20px;
  position: absolute;
}

.petc-attr-editor input {
  background-color: var(--editor-bg);
  border-radius: 5px;
  border: 1px solid rgba(210, 210, 210, 1);
  color: grey;
  font-family: var(--font);
  font-size: larger;
  font-weight: bold;
  text-align: center;
}

.petc-attr-editor td {
  text-align: center;
}

.petc-attr-editor option,
.petc-attr-editor select {
  background-color: var(--editor-bg);
  border-radius: 5px;
  border: 1px solid rgba(210, 210, 210, 1);
  color: grey;
  font-family: var(--font);
  font-size: larger;
  font-weight: bold;
  text-align: center;
  text-indent: 30%;
  width: 100%;

  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

