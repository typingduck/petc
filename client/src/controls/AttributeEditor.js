import React from 'react'
import Modal from './Modal'

import './AttributeEditor.css'
import {DEFAULT_CLASS_NAME} from '../model/model'

/**
 * Modal to edit label/class of edges or nodes.
 * A temporary solution using a dialog. Would be much better to edit in place
 * to keep the look and feel of a whiteboard but that will require a lot
 * more client side code to enable the feeling.
 */
class AttributeEditor extends React.Component {
  constructor (props) {
    super(props)

    this.state = {

      openEditor: false,
      label: null,
      className: null,
      classes: []
    }

    this.update = this.update.bind(this)
    this.apply = this.apply.bind(this)
    this.cancel = this.cancel.bind(this)
  }

  componentWillReceiveProps (nextProps) {
    const editAttributes = nextProps.controls.editAttributes
    const openEditor = Boolean(editAttributes)
    if (openEditor) {
      const itemId = nextProps.controls.editAttributes.id
      let toEdit, classes
      if (nextProps.doc.nodes[itemId]) {
        toEdit = nextProps.doc.nodes[itemId]
        classes = Object.keys(nextProps.doc.style.nodes)
      } else if (nextProps.doc.edges[itemId]) {
        toEdit = nextProps.doc.edges[itemId]
        classes = Object.keys(nextProps.doc.style.edges)
      } else {
        return
      }
      if (!classes.includes(DEFAULT_CLASS_NAME)) {
        classes = [DEFAULT_CLASS_NAME].concat(classes)
      }
      this.setState({
        openEditor: openEditor,
        label: toEdit.label || '',
        classes: classes,
        className: toEdit.className || DEFAULT_CLASS_NAME,
        pageX: editAttributes.pageX + 5,
        pageY: editAttributes.pageY + 5
      })
    } else {
      this.setState({ openEditor: false })
    }
  }

  update (e) {
    this._src = e.updated_src
  }

  onLabelChange (value) {
    this.setState({label: value})
  }

  onClassChange (value) {
    this.setState({className: value})
  }

  apply () {
    const itemId = this.props.controls.editAttributes.id
    if (getNode(this.props, itemId)) {
      const node = Object.assign({}, getNode(this.props, itemId))
      node.label = this.state.label
      node.className = this.state.className
      this.props.updateNode(node)
    } else if (getEdge(this.props, itemId)) {
      const edge = Object.assign({}, getEdge(this.props, itemId))
      edge.label = this.state.label
      edge.className = this.state.className
      this.props.updateEdge(edge)
    }
    this.cancel()
  }

  cancelNode () {
    this.setState({label: null, className: null})
    this.props.onCancel()
  }

  cancel () {
    this.props.setEditAttributes(null)
  }

  render () {
    return (
      <Modal
        className='petc-attr-editor'
        isOpen={this.state.openEditor}
        onRequestClose={this.cancel}
        pageX={this.state.pageX}
        pageY={this.state.pageY}
      >
        <table><tbody>
          <tr>
            <td>label:</td>
            <td>
              <input
                type='text'
                onChange={(ev) => this.onLabelChange(ev.target.value)}
                value={this.state.label}
              />
            </td>
          </tr>
          <tr>
            <td>class:</td>
            <td>
              <select
                value={this.state.className}
                onChange={(ev) => this.onClassChange(ev.target.value)}
              >
                {this.state.classes.map(className =>
                  <option key={className}>{className}</option>
                )}
              </select>
            </td>
          </tr>
        </tbody></table>
        <div style={{textAlign: 'center'}}>
          <button onClick={this.apply}>apply</button>
          <button onClick={this.cancel}>cancel</button>
        </div>
      </Modal>
    )
  }
}

function getNode (props, nodeId) {
  return props.doc.nodes[nodeId]
}

function getEdge (props, edgeId) {
  return props.doc.edges[edgeId]
}

export default AttributeEditor
