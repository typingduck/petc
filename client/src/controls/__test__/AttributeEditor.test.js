/* global beforeEach expect it jest */
import React from 'react'
import Renderer from 'react-test-renderer'

import AttributeEditor from '../AttributeEditor'

const props = {}

const nodeProps = {
  doc: {
    nodes: {
      'anode': { id: 'anode', className: 'node style a' }
    },
    edges: {
      'anedge': { id: 'anedge', className: 'edge style a' }
    },
    style: {
      nodes: {
        'node style a': {},
        'node style b': {}
      },
      edges: {
        'edge style a': {},
        'edge style b': {}
      }
    }
  },
  controls: {
    editAttributes: { id: 'anode', pageX: 95, pageY: 95 }
  },
  setEditAttributes: jest.fn(),
  updateNode: jest.fn(),
  updateEdge: jest.fn()
}

beforeEach(() => {
  jest.clearAllMocks()
})

it('hides attribute editor by default', () => {
  const component = Renderer.create(<AttributeEditor {...props} />)
  expect(component.toJSON()).toBeNull()
})

it('renders attribute editor for node', () => {
  const component = Renderer.create(<AttributeEditor {...props} />)
  component.getInstance().componentWillReceiveProps(nodeProps)
  expect(component.toJSON()).not.toBeNull()
  expect(component.toJSON()).toMatchSnapshot()
})

it('calls reset edit attributes element on cancel', () => {
  const component = Renderer.create(<AttributeEditor {...nodeProps} />)
  component.getInstance().componentWillReceiveProps(nodeProps)

  const cancelButton = component.root.findAllByType('button')[1]
  cancelButton.props.onClick()
  expect(nodeProps.setEditAttributes).toHaveBeenCalledWith(null)
})

it('fills label on apply button press', () => {
  const component = Renderer.create(<AttributeEditor {...nodeProps} />)
  component.getInstance().componentWillReceiveProps(nodeProps)

  fillLabelInput(component, 'new label')

  const applyButton = component.root.findAllByType('button')[0]
  applyButton.props.onClick()

  expect(nodeProps.updateNode).toHaveBeenCalledWith({
    id: 'anode', label: 'new label', className: 'node style a'
  })
})

it('fills class on apply button press', () => {
  const component = Renderer.create(<AttributeEditor {...nodeProps} />)
  component.getInstance().componentWillReceiveProps(nodeProps)

  fillClassInput(component, 'style b')

  const applyButton = component.root.findAllByType('button')[0]
  applyButton.props.onClick()

  expect(nodeProps.updateNode).toHaveBeenCalledWith({
    id: 'anode', label: '', className: 'style b'
  })
})

const edgeProps = Object.assign({}, nodeProps,
  { controls: { editAttributes: { id: 'anedge', pageX: 95, pageY: 95 } } }
)

it('renders attribute editor for edge', () => {
  const component = Renderer.create(<AttributeEditor {...props} />)
  component.getInstance().componentWillReceiveProps(edgeProps)
  expect(component.toJSON()).not.toBeNull()
  expect(component.toJSON()).toMatchSnapshot()
})

it('updates edge on apply', () => {
  const component = Renderer.create(<AttributeEditor {...edgeProps} />)
  component.getInstance().componentWillReceiveProps(edgeProps)

  fillLabelInput(component, 'new edge label')
  fillClassInput(component, 'edge style b')

  const applyButton = component.root.findAllByType('button')[0]
  applyButton.props.onClick()

  expect(nodeProps.updateEdge).toHaveBeenCalledWith({
    id: 'anedge', label: 'new edge label', className: 'edge style b'
  })
})

function fillLabelInput (component, value) {
  const labelField = component.root.findByType('input')
  labelField.props.onChange({ target: { value } })
}

function fillClassInput (component, value) {
  const classField = component.root.findByType('select')
  classField.props.onChange({ target: { value } })
}
