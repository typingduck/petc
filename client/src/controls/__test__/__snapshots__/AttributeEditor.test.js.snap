// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders attribute editor for edge 1`] = `
<div
  className="petc-modal-overlay"
  onClick={[Function]}
>
  <div
    className="petc-modal petc-attr-editor"
    onClick={[Function]}
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  >
    <table>
      <tbody>
        <tr>
          <td>
            label:
          </td>
          <td>
            <input
              onChange={[Function]}
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td>
            style:
          </td>
          <td>
            <select
              onChange={[Function]}
              value="edge style a"
            >
              <option>
                default
              </option>
              <option>
                edge style a
              </option>
              <option>
                edge style b
              </option>
            </select>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        apply
      </button>
      <button
        onClick={[Function]}
      >
        cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`renders attribute editor for node 1`] = `
<div
  className="petc-modal-overlay"
  onClick={[Function]}
>
  <div
    className="petc-modal petc-attr-editor"
    onClick={[Function]}
    style={
      Object {
        "left": 100,
        "top": 100,
      }
    }
  >
    <table>
      <tbody>
        <tr>
          <td>
            label:
          </td>
          <td>
            <input
              onChange={[Function]}
              type="text"
              value=""
            />
          </td>
        </tr>
        <tr>
          <td>
            style:
          </td>
          <td>
            <select
              onChange={[Function]}
              value="node style a"
            >
              <option>
                default
              </option>
              <option>
                node style a
              </option>
              <option>
                node style b
              </option>
            </select>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <button
        onClick={[Function]}
      >
        apply
      </button>
      <button
        onClick={[Function]}
      >
        cancel
      </button>
    </div>
  </div>
</div>
`;
